File: cpp/day_22b.cpp
#include <algorithm>
#include <deque>
#include <fstream>
#include <functional>
#include <iostream>
#include <string>
#include <set>
#include <map>
#include <vector>
// TODO(vss): improve the previous_rounds so it can be generalized to n players. Rest already generalized. Multimap?
std::tuple<int, std::vector<std::deque<int>>, std::vector<int>> PlayRecursiceCombat(std::vector<std::deque<int>> player_decks, std::vector<int> players) {
  std::map<std::deque<int>, std::set<std::deque<int>>> previous_rounds;
  while (player_decks.size() > 1) {
    if(auto it = previous_rounds.find(player_decks[0]);
       it != previous_rounds.end() && it->second.find(player_decks[1]) != it->second.end()) {
        return {1, player_decks, players}; // player 1 wins game
    } else {
      previous_rounds[player_decks[0]].insert(player_decks[1]);
    }

    size_t winning_player = -1;
    int hightest_card = -1;
    std::vector<int> top_cards(player_decks.size());
    bool n_remaining_greater_than_card_value = true;
    for(int i = 0; i < player_decks.size(); i++) {
      if(hightest_card < player_decks[i].front() ) {
        hightest_card = player_decks[i].front();
        winning_player = players[i];
      }
      if(player_decks[i].front() > player_decks[i].size() - 1) {
        n_remaining_greater_than_card_value = false;
      }
      top_cards[i] = player_decks[i].front();
      player_decks[i].pop_front();
    }

    if(n_remaining_greater_than_card_value) {
      std::vector<std::deque<int>> subgame_player_decks;
      subgame_player_decks.reserve(top_cards.size());
      for(int i = 0; i < top_cards.size(); i++) {
        subgame_player_decks.emplace_back(std::begin(player_decks[i]), std::begin(player_decks[i]) + top_cards[i]);
      }
      auto result = PlayRecursiceCombat(subgame_player_decks, players);
      winning_player = std::get<0>(result);
    }
    auto dist = std::distance(std::begin(players), std::find(std::begin(players), std::end(players), winning_player));
    std::rotate(std::begin(top_cards), std::begin(top_cards) + dist, std::end(top_cards));

    if(std::find(std::begin(players), std::end(players), winning_player) == std::end(players)) {
      players.emplace_back(winning_player);
      player_decks.emplace_back();
    }

    for(const auto& top_card : top_cards) {
      auto dist = std::distance(std::begin(players), std::find(std::begin(players), std::end(players), winning_player));
      player_decks[dist].push_back(top_card);
    }

    for(int i = 0; i < player_decks.size(); i++) {
      if(player_decks[i].empty()) {
        player_decks.erase(std::begin(player_decks) + i);
        players.erase(std::begin(players) + i);
        i--;
      }
    }
  }
  return {players[0], player_decks, players};
}

int main() {
  std::fstream file{"../input/day_22_input"};
  std::string line;
  std::vector<std::deque<int>> player_decks;
  std::vector<int> players;
  while(std::getline(file, line)) {
    line.erase(std::remove_if(std::begin(line), std::end(line), [](const char c) {return !isprint(c);}), std::end(line));
    if(line == "") continue;
    if(line.substr(0, 6) == "Player") {
      player_decks.emplace_back();
      players.push_back(std::stoi(line.substr(7, line.size() - 8)));
    } else {
      player_decks.back().push_back(std::stoi(line));
    }
  }
  int total_players = players.size();

  // Play
  auto result = PlayRecursiceCombat(player_decks, players);
  auto dist = std::distance(std::begin(std::get<2>(result)),
    std::find(std::begin(std::get<2>(result)), std::end(std::get<2>(result)), std::get<0>(result)));
  long long score = 0;
  const auto& deck = std::get<1>(result)[dist];
  for(int i = 0; i < deck.size(); i++) {
    score += ((i+1) * deck[deck.size() - 1 - i]);
  }
  std::cout << score << '\n';
  return score;
}

File: python/day_22b.py
#!/usr/bin/python3
import copy

def PlayRecursiveCombat(players, player_decks):
    previous_rounds = list()
    while len(player_decks) > 1:
        for i in range(0, len(previous_rounds)):
            if player_decks[0] == previous_rounds[i][0] and player_decks[1] == previous_rounds[i][1]:
                return 1, player_decks, players
        else:
            previous_rounds.append([copy.deepcopy(player_decks[0]),  copy.deepcopy(player_decks[1])])
        winning_player = -1
        hightest_card = -1
        top_cards = list()
        n_remaining_greater_than_card_value = True
        for i in range(0, len(player_decks)):
            if hightest_card < player_decks[i][0]:
                hightest_card = player_decks[i][0]
                winning_player = players[i]
            if player_decks[i][0] > len(player_decks[i]) - 1:
                n_remaining_greater_than_card_value = False;
            top_cards.append(player_decks[i].pop(0))

        if n_remaining_greater_than_card_value:
            subgame_player_decks = list()
            for i in range(0, len(top_cards)):
                subgame_player_decks.append(player_decks[i][0:top_cards[i]]);
            result = PlayRecursiveCombat(copy.deepcopy(players), subgame_player_decks)
            winning_player = result[0]
        for i in range(0, len(players)):
            if players[i] != winning_player:
                top_cards.append(top_cards.pop(0))
            else:
                for top_card in top_cards:
                    player_decks[i].append(top_card)
                break
        indices_to_remove = list()
        for i in range(0, len(player_decks)):
            if len(player_decks[i]) == 0:
                indices_to_remove.append(i)
        for i in range(0, len(indices_to_remove)):
            player_decks.pop(indices_to_remove[len(indices_to_remove) - i - 1])
            players.pop(indices_to_remove[len(indices_to_remove) - i - 1])
    return players[0], player_decks, players

def main():
    file = open("../input/day_22_input")
    players = list()
    player_decks = list()
    for line in file:
        line = line.strip('\n')
        if line == "":
            continue
        if line[0:6] == "Player":
            player_decks.append([])
            players.append(int(line[7:len(line)-1]))
        else:
            player_decks[-1].append(int(line))
    total_players = len(players)
    result = PlayRecursiveCombat(players, player_decks)
    winning_player, player_decks, players = result
    player_deck = list()
    score = 0
    for i in range(0, len(players)):
        if players[i] == winning_player:
            player_deck = player_decks[i]
    for i in range(0, len(player_deck)):
        score = score + ((i+1) * player_deck[len(player_deck) - 1 - i]);
    print(score)
    return score

if __name__ == "__main__":
    main()
